File: app/src/main/java/com/cabbage/scaffold/MainActivity.java
package com.cabbage.scaffold;

import android.Manifest;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import com.jakewharton.rxbinding2.view.RxView;
import com.tbruyelle.rxpermissions2.RxPermissions;

import butterknife.BindView;
import butterknife.ButterKnife;
import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.annotations.NonNull;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Consumer;
import timber.log.Timber;

public class MainActivity extends AppCompatActivity {

    @BindView(R.id.toolbar) Toolbar mToolbar;
    @BindView(R.id.fab) FloatingActionButton mFab;

    private CompositeDisposable compositeDisposable;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);
        setSupportActionBar(mToolbar);

        compositeDisposable = new CompositeDisposable();
        RxPermissions rxPermissions = new RxPermissions(this);
        rxPermissions.setLogging(true);
        String locationPermission = Manifest.permission.ACCESS_FINE_LOCATION;

        Observable.merge(rxPermissions.request(locationPermission),
                RxView.clicks(mFab).compose(rxPermissions.ensure(locationPermission)))
                .subscribe(new Consumer<Boolean>() {
                    @Override
                    public void accept(@NonNull Boolean aBoolean) throws Exception {
                        Toast.makeText(MainActivity.this, aBoolean.toString(), Toast.LENGTH_SHORT).show();
                    }
                });
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (compositeDisposable != null && !compositeDisposable.isDisposed()) {
            compositeDisposable.dispose();
        }
    }
}

File: app/src/main/java/com/cabbage/scaffold/MainActivity.kt
package com.cabbage.scaffold

import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.Toolbar
import android.widget.Toast
import butterknife.BindView
import butterknife.ButterKnife
import com.cabbage.scaffold.mvp.MvpView
import com.jakewharton.rxbinding2.view.RxView
import com.tbruyelle.rxpermissions2.RxPermissions
import io.reactivex.rxkotlin.subscribeBy

class MainActivity : AppCompatActivity(), MvpView {

    var presenter: MainPresenter? = null

    @BindView(R.id.toolbar) lateinit var mToolbar: Toolbar
    @BindView(R.id.fab) lateinit var mFab: FloatingActionButton

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        ButterKnife.bind(this)
        setSupportActionBar(mToolbar)

        RxView.clicks(mFab).subscribeBy(onNext = { presenter?.ensureLocationPermission() })
    }

    public override fun onStart() {
        super.onStart()
        if (presenter == null) presenter = MainPresenter(RxPermissions(this))
        presenter?.attachView(this)
        presenter?.ensureLocationPermission()
    }

    public override fun onStop() {
        super.onStop()
        presenter?.detachView()
    }

    fun showLocationPermissionResult(granted: Boolean) {
        Toast.makeText(this, "Granted: $granted", Toast.LENGTH_SHORT).show()
    }
}
